[bitnami/keycloak] switches keycloak-metrics service to poin to new port

maschmi opened the pull request:
<!--
 Before you open the request please review the following guidelines and tips to help it be more easily integrated:

 - Describe the scope of your change - i.e. what the change does.
 - Describe any known limitations with your change.
 - Please run any tests or examples that can exercise your modified code.

 Thank you for contributing! We will try to test and integrate the change as soon as we can, but be aware we have many GitHub repositories to manage and can't immediately respond to every request. There is no need to bump or check in on a pull request (it will clutter the discussion of the request).

 Also don't be worried if the request is closed or not integrated sometimes the priorities of Bitnami might not match the priorities of the pull request. Don't fret, the open source community thrives on forks and GitHub makes it easy to keep your changes in a forked repo.
 -->

### Description of the change

This change adds a new configuration value to configure the containers metrics port. This is helpflus as with Keycloak 25 the metrics port was changed from 8080 to 9000. Enabling metrics will now lead to a correctly adapted network policy and also pointing the keycloak-metric service to this port. The keycloak-metric service is still exposed at port 8080.

### Benefits

Keycloak must not be started with KC_LEGACY_OBSERVABILITY_INTERFACE environment variable set. Therefore, it does not produce a deprecation warning when starting.

### Possible drawbacks

If a user has set the KC_LEGACY_OBSERVABILITY_INTERFACE in the keycloak environment variables, which restored metrics by exposing them on port 8080, they now either need to restart keycloak without the legacy behaviour or setting the containerPorts.metrics value to 8080. There is a note in the README wehre containerPorts.metrics is described.

### Applicable issues

- fixes #28630 

### Additional information



### Checklist

<!-- [Place an '[X]' (no spaces) in all applicable fields. Please remove unrelated fields.] -->

- [X] Chart version bumped in `Chart.yaml` according to [semver](http://semver.org/). This is *not necessary* when the changes only affect README.md files.
- [X] Variables are documented in the values.yaml and added to the `README.md` using [readme-generator-for-helm](https://github.com/bitnami/readme-generator-for-helm)
- [X] Title of the pull request follows this pattern [bitnami/<name_of_the_chart>] Descriptive title
- [X] All commits signed off and in agreement of [Developer Certificate of Origin (DCO)](https://github.com/bitnami/charts/blob/main/CONTRIBUTING.md#sign-your-work)

fmulero commented on bitnami/keycloak/Chart.yaml:
```diff
@@ -33,4 +33,4 @@ maintainers:
 sources:
 - https://github.com/bitnami/charts/tree/main/bitnami/keycloak
-version: 24.0.4
+version: 24.0.5
```
We are adding a new feature, users can now change the metrics port, according to [semver](https://semever.org) we should bump the minor version:

```suggestion
version: 24.1.0
```

fmulero reviewed (commented):
Thanks a lot @maschmi for your contribution and taking care of this issue.

I've left some suggestions, little things. Could you give them a glance?

I also think the change in the metrics service, setting the default service port to 9000, would be worth. I think this is the perfect moment to change it.

maschmi commented on bitnami/keycloak/templates/statefulset.yaml:
```diff
@@ -249,6 +249,11 @@ spec:
               containerPort: {{ .Values.containerPorts.https }}
               protocol: TCP
             {{- end }}
+            {{- if and (.Values.metrics.enabled) (not (eq (.Values.containerPorts.http | quote) (.Values.containerPorts.metrics | quote) )) }}
```
When removing the 'quote' calling `helm template --set metrics.enabled=true --set containerPorts.metrics=8080 .` fails as well as calling `helm install` with the same parameters. (My helm Version: `version.BuildInfo{Version:"v3.16.2", GitCommit:"13654a52f7c70a143b1dd51416d633e1071faffb", GitTreeState:"", GoVersion:"go1.23.2"}`)


```
helm template --debug --set metrics.enabled=true --set containerPorts.metrics=8080 .                                                 1 ↵
install.go:224: 2024-11-04 20:27:38.309844737 +0100 CET m=+0.023248418 [debug] Original chart version: ""
install.go:241: 2024-11-04 20:27:38.309873812 +0100 CET m=+0.023277483 [debug] CHART PATH: /home/martin/src/bitnami_charts/bitnami/keycloak


Error: template: keycloak/templates/networkpolicy.yaml:69:51: executing "keycloak/templates/networkpolicy.yaml" at <eq (.Values.containerPorts.http) (.Values.containerPorts.metrics)>: error calling eq: incompatible types for comparison
helm.go:86: 2024-11-04 20:27:38.341824604 +0100 CET m=+0.055228285 [debug] template: keycloak/templates/networkpolicy.yaml:69:51: executing "keycloak/templates/networkpolicy.yaml" at <eq (.Values.containerPorts.http) (.Values.containerPorts.metrics)>: error calling eq: incompatible types for comparison
```
When using the Values file only the `quote` does not matter. Therefore, I did not remove the quote. However, I'm not a helm expert and may be doing things wrong here. Also, if this is intended behaviour I'm happily adding a commit removing the quote.

fmulero replied:
I've just reproduce the issue you mentioned. `--set` parameter is parsing the number as float and the eq function can not compare an integer against a float. You can workaorund the issue with `--set-json` but I think is clever something like this:
```suggestion
            {{- if and (.Values.metrics.enabled) (not (eq (.Values.containerPorts.http | int) (.Values.containerPorts.metrics | int) )) }}
```

This will ensure that we have set integers in the ports

maschmi commented on bitnami/keycloak/templates/networkpolicy.yaml:
```diff
@@ -66,6 +66,9 @@ spec:
     - ports:
         {{- /* Constant in code: https://github.com/keycloak/keycloak/blob/ce8e925c1ad9bf7a3180d1496e181aeea0ab5f8a/operator/src/main/java/org/keycloak/operator/Constants.java#L60 */}}
         - port: 7800
+        {{- if and (.Values.metrics.enabled) (not (eq (.Values.containerPorts.http | quote) (.Values.containerPorts.metrics | quote) )) }}
```
see other `quote` suggestion

maschmi commented:
I've tested and added the suggestions you had. Except the one with removing the `quote`, I detailed the reason in a comment.
I've also moved the metrics service to listen on port `9000`. This may however break some existing things, which was the reason I did not change it. May I ask the question if there is a specific need to have the metrics service listening on the same port as the metrics endpoint in the container?

fmulero reviewed (commented):
Thanks a lot @maschmi, I've just left a comment a about the issue you've found with the types.

About changing the default value for `metrics.service.ports.http`, I agree, it could lead into errors but from Keycloak 25.0.0 the default port for the metrics is the the same as the management interface, [the 9000](https://www.keycloak.org/docs/latest/release_notes/index.html#management-port-for-metrics-and-health-endpoints), so I think we should use the same values as the upstream project. Anyway the users could set the port and use again 8080. It would be great if you could add a note for this version here: https://github.com/bitnami/charts/blob/528d26f262befe7263b0f0e7728dbe706dd33c92/bitnami/keycloak/README.md?plain=1#L682-L685

maschmi commented:
@fmulero I've added both changed. Thank you for your guidance.

fmulero reviewed (approved):
Great @maschmi !!

LGTM
